### ISSUE TITLE
[REQUEST] Apple Silicon GPU? 🥺

### ISSUE THREAD

**Issue by aidan-gibson** [aristocratos/btop#701]
https://github.com/tlkh/asitop does this via [powermetrics](https://www.unix.com/man-page/osx/1/powermetrics/) (sudo required). I don't mind the sudo requirement, it seems unavoidable. 

This seemed like a viable sudoless alternative (https://github.com/op06072/NeoAsitop, which relies on https://github.com/dehydratedpotato/socpowerbud) but neither worked on my machine and it seems to be a widespread issue.

**Comment by martinoturrina** [comment]
This would be so good!

**Comment by emanuelez** [comment]
I have an implementation of this here: https://github.com/emanuelez/btop/tree/apple-silicon

It was LLM generated and I don't think it's fair to push for review some code that I don't fully understand. I was simply scratching an itch and this works for me so anybody please feel free to:

- review the code and make a proper PR. I don't expect any credit or attribution really.
- build it and run it. It works quite nicely for me but it does require sudo

**Comment by martinoturrina** [comment]
> * but it does require sudo

Seems like macmon does most of the things needed without sudo, so maybe that could be interesting to take a look to see if it's possible to avoid requiring sudo for btop also

**Comment by emanuelez** [comment]
Well spotted! Here's a sudo-less version that comes with the same disclaimers as my previous comment: https://github.com/emanuelez/btop/tree/gpu-apple-silicon-nosudo

<img width="1506" height="381" alt="Image" src="https://github.com/user-attachments/assets/1370fbcb-8202-48d4-8901-4ec280089dcb" />

**Comment by martinoturrina** [comment]
To me the code looks ok, but I'm also not an expert. 

One concern is that usually GPU on linux is shown under the CPU cores in a more conceived way, the screen you posted is a bit different layout which is not bad but overall maybe is best to keep the standard layout

<img width="1252" height="224" alt="Image" src="https://github.com/user-attachments/assets/610abcb3-e078-4580-a4dd-a3e778a387b0" />

**Comment by emanuelez** [comment]
I was just pressing 5 to show more of the GPU stats :)

<img width="1506" height="293" alt="Image" src="https://github.com/user-attachments/assets/dbce38d1-1491-4a3c-8b2f-280b5bf1df57" />

**Comment by martinoturrina** [comment]
I think you can submit the PR and mark it as AI generated, I think is a good contribution even if the code was generated

**Comment by Ruaneri-Portela** [comment]
I ended up duplicating the feature because I didn't see yours, my mistake.
 
I noticed that you are missing some information here.
I ended up implementing it without seeing your implementation, but I added the view via IOkit.

<img width="2527" height="286" alt="Image" src="https://github.com/user-attachments/assets/32180b07-680b-4b86-ac1e-0e657c74ef36" />

With IOKit, you can see both reserved and used memory.
I ended up using Powermetrics too, but your implementation is nicer.

However, if we go beyond the scope of the GPU, we can reuse the clock and power usage data for the CPU.

<img width="2538" height="277" alt="Image" src="https://github.com/user-attachments/assets/56a92c5c-e33c-4c05-83a8-f5521c9fca73" />

[emanuelez](https://github.com/emanuelez) managed to do it without sudo, he looked at the IOReports. I only looked at the PerformanceStatistics.

<img width="976" height="847" alt="Image" src="https://github.com/user-attachments/assets/13b55d1c-f530-4470-8fbd-3c320c85c0ad" />

**Comment by marcoxyz123** [comment]
Hi All

Without Sudo:

CPU Temp fixed and added Efficient Cores and Powercores Detection, GPU  count, pwr, tmp, ANE (Apple Neural Engine) count, per, cmd/s

<img width="3150" height="720" alt="Image" src="https://github.com/user-attachments/assets/bd7f1454-d7b2-4cc8-93de-91f40755b292" />

GPU panel on 5, Split View for ANE with 6, (it is unlikely that Apple has more than one GPU group ;-) )

 
<img width="3104" height="1306" alt="Image" src="https://github.com/user-attachments/assets/06222150-79c6-4490-a4e3-5087350eab7d" />

<img width="3126" height="1296" alt="Image" src="https://github.com/user-attachments/assets/e8c3d273-312d-404a-93fe-d7b9d22f6618" />

Power panel on  7, with CPU power & Temp, GPU Power and Temp, ANE Power. Temp and  diapers from cpu and gpu panel

 
<img width="3124" height="1688" alt="Image" src="https://github.com/user-attachments/assets/b4d2df66-c2ff-4bc2-9a06-ece816afa1c6" />

<img width="3174" height="1760" alt="Image" src="https://github.com/user-attachments/assets/9fc38c40-6b7c-4af5-b600-66b63c686835" />

And All together:

 
<img width="3208" height="2326" alt="Image" src="https://github.com/user-attachments/assets/9066a115-dd5b-4b79-bef7-c518f9e291de" />

sensors.cpp gefixt fuer Apple Silicon.

Testing now for 72 hours.

Greetings Marco

**Comment by eza-captureone** [comment]
Good job @marcoxyz123 and @Ruaneri-Portela.

You can dissect and use my code in any way you prefer if it helps your implementation in any way :)

**Comment by Ruaneri-Portela** [comment]
@marcoxyz123 and @emanuelez 

I think we could combine our implementations.
The framework accessed data from the NPU, I put the data in memory.

I think we can keep the power metrics with fallback in case they change the API? And also complement it with the CPU clock... I don't know if we can find that within IOkit.

In the IOkit API, you can get GPU usage by process. Would it be possible to create a new field in processes to put its usage?

<img width="997" height="167" alt="Image" src="https://github.com/user-attachments/assets/3c5a64dd-276c-4105-8a49-1ff313ba982d" />

The AGXDeviceUserClient children of AGXAccelerator contain the process pid, a field called AppUsage with the accumulated GPU time.

But I don't know how to convert this to % device usage. 

I'll make an extractor for this information. If anyone there knows how to convert it and then how to put it in the processes section

EDIT:

<img width="1342" height="400" alt="Image" src="https://github.com/user-attachments/assets/af394a8f-677e-4abe-b973-ed77f857f676" />

I did the part to look up the processes, but I don't know how to convert them into actual percentages.

EDIT2:

With the help of AI, it helped me adjust based on the total % value, and the values matched closely with the Activity Monitor.

<img width="1041" height="489" alt="Image" src="https://github.com/user-attachments/assets/30ddf00a-c1e0-4797-a16a-1170dc80cffc" />

**Comment by Ruaneri-Portela** [comment]
I took the code written by @emanuelez and adapted it to my code, making some changes. I moved some iokit definitions to a new file to make it reusable between files. 

I left the Powermetrics backend, which provides the CPU clock and power, but we can reuse the GPU code to find CPU data such as watts and frequency without Powermetrics.

It would be nice to move IOReport out of the GPU, put it in Shared, and record what can be read from it. I added a filter that identifies the GPU driver and reads only that one. I'm trying to make it compatible to iterate on IOAccelerator. On Mac Intel, there may be more than one GPU. However, I have no way to test it! 

<img width="2334" height="1592" alt="Image" src="https://github.com/user-attachments/assets/f9edc550-801b-4cbc-a7e4-8400d3174bd5" />

I wanted to see @marcoxyz123's additions before making a PR.

But he didn't leave a public commit on his git. If you don't mind sharing it with us.

My changes are in my fork.

**Comment by marcoxyz123** [comment]
I will do that later, had allot of changes today, working now on the cpu/gpu panel to work correct with the new per panel

<img width="3194" height="2076" alt="Image" src="https://github.com/user-attachments/assets/d47838e0-b2b7-4f18-afe7-4d16a33769f2" />

Ciao Marco

**Comment by marcoxyz123** [comment]
@emanuelez  Do you have already an PR? Or should I include your change in my PR?

**Comment by marcoxyz123** [comment]
@Ruaneri-Portela You need to add later the vram part because I cannot test it because I have only M3,4 here.

**Comment by Ruaneri-Portela** [comment]
@marcoxyz123 I've already brought his changes to my fork, if you don't mind can you do a PR on my fork?

The one I have to test the vram is an M4 Pro, I think it's similar to yours

I think we can make a change if there is ANE we put it on 6 and the power on 5 then we leave it bigger than 6 for GPU when there is no Ane, we put power on 5 and the gpu bigger than 6. I think about leaving it prepared for if someone wants to leave it ok on Intel Macs

**Comment by emanuelez** [comment]
> [@emanuelez](https://github.com/emanuelez) Do you have already an PR? Or should I include your change in my PR?

I didn't push any PR. Feel free to use my code in any way you prefer and push a PR :)

**Comment by marcoxyz123** [comment]
@emanuelez Will do in the next two days here the finished thing:

CPU alone showing GPU and ANE

<img width="3214" height="1278" alt="Image" src="https://github.com/user-attachments/assets/dc4a622a-4b3a-4feb-978c-ba1c09da5db9" />

CPU and POWER no redundancy for PWR and TEMP

<img width="3228" height="1246" alt="Image" src="https://github.com/user-attachments/assets/a0aa684c-b77b-4559-ab5d-4257f17b1bea" />

CPU / POWER and GPU Panel no redundancy for PWR and TEMP

<img width="3196" height="1270" alt="Image" src="https://github.com/user-attachments/assets/4914f1f2-45bf-498c-9ebb-802f13eb3171" />

CPU and GPU with PWR and TEMP

<img width="3220" height="1280" alt="Image" src="https://github.com/user-attachments/assets/440cfca9-5f84-4473-b101-85607f12c92d" />

All together with the new option net above proc and right of mem (without proc right of mem, pass "3" twice ;-) )

<img width="3220" height="2142" alt="Image" src="https://github.com/user-attachments/assets/807fa34b-9367-408b-8958-7e51509b34e4" />

All together but Proc exclusive on the bottom press twice "4"

<img width="3230" height="2744" alt="Image" src="https://github.com/user-attachments/assets/15456671-0124-4f62-af04-6d8aed1a1b60" />

For Intel Macs ANE can be replaced than with the VRAM information by @Ruaneri-Portela. I let bebop running now. will see how stable it is for two days.

Ciao Marco

**Comment by marcoxyz123** [comment]
btop not bebop LOL. Additional is fixed the disks view is showing internal HDDs, external HDD with connection Thunderbolt/USB, Network Drives with AFS,SMB and NFS, Images with DMG, ISO and IMG. In additional the men panel is space optimised to fill the whole panel. The disk part has for smaller vies now scrolling enabled. Default is still the proc panel but "tab" gives the disk panel the focus so that you can scroll the your HDD even in smaller views.

Before:

<img width="776" height="1472" alt="Image" src="https://github.com/user-attachments/assets/0d269142-0dcb-48b1-9113-e8a6b03d2ac3" />

After:

<img width="1594" height="822" alt="Image" src="https://github.com/user-attachments/assets/f079bc1e-8434-4a68-bd4a-4cc2b3de206e" />

No sleeping Ciao

**Comment by Ruaneri-Portela** [comment]
@marcoxyz123 Can you make a change? In the field of processes put the use of GPU?

I've already done the part that extracts this data

**Comment by Ruaneri-Portela** [comment]
About the use of ram, I would like to keep it even on Apple silicon, the IOkit says the current memory use by the gpu and how much of the total memory is reserved for the gpu

**Comment by marcoxyz123** [comment]
Yes for sure I can do this tomorrow, I build in a bug to if the window gets to small, the error dialog with the information that the window is to small is not coming up anymore. I have to fix this to because of the new panel ordering. BTW the ANE graph in GPU can be switched by pressing "6".

**Comment by Ruaneri-Portela** [comment]
I moved the calculation of % usage per process into my GPU class. An unordered map with the process pid as the key for quick lookup.

<img width="1200" height="948" alt="Image" src="https://github.com/user-attachments/assets/3d37ace4-1b0b-4064-a39c-a265d7fef6a0" />

The only thing I wasn't able to get was the temperature. If anyone out there knows how to do it, please let me know. I copied @emanuelez 's method, but it didn't work for me.

**Comment by marcoxyz123** [comment]
So stop stops working after 5 hours with:

===> btop++ v1.4.6
2026-01-13 (02:19:45) | ERROR: Stall in Runner thread, restarting!
2026-01-13 (07:09:54) | WARNING: Failed to join _runner thread on exit!

I let now implement this fix:

This can happen because of IOReportCreateSamples()  can block indefinitely. This can also happen on high cpu load:

1. Async IOReport with Timeout (Primary Fix)

// Use GCD dispatch queue with timeout for IOReport calls
dispatch_semaphore_t sem = dispatch_semaphore_create(0);
__block CFDictionaryRef result = nullptr;

dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0), ^{
    result = IOReportCreateSamples(subscription, channels, nullptr);
    dispatch_semaphore_signal(sem);
});

// Wait max 2 seconds, use cached values if timeout
if (dispatch_semaphore_wait(sem, dispatch_time(DISPATCH_TIME_NOW, 2 * NSEC_PER_SEC)) != 0) {
    Logger::warning("IOReport timeout, using cached values");
    return cached_metrics;  // Graceful degradation
}
2. Adaptive Timeout (Secondary)

// Adjust runner timeout based on system load
int get_adaptive_timeout() {
    double load = get_system_load();
    if (load > 8.0) return 15000;      // 15s for very high load
    else if (load > 4.0) return 10000; // 10s for high load
    return 5000;                        // 5s default
}
3. Graceful Skip Instead of Restart

// Don't restart thread, just skip cycle
if (active) {
    Logger::warning("Collection slow, skipping cycle");
    active = false;  // Release lock
    continue;        // Skip to next iteration
    // Don't restart thread!
}
4. Cached Metrics Fallback

struct CachedMetrics {
    AppleSiliconGpuMetrics gpu;
    std::chrono::steady_clock::time_point timestamp;
    bool valid = false;
};
static CachedMetrics cache;

// Use cache if fresh enough (< 5 seconds old)
if (timeout_occurred && cache.valid && 
    age(cache.timestamp) < 5s) {
    return cache.gpu;
}

Will See if this fix works. Testing the next hours.

Ciao Marco

**Comment by marcoxyz123** [comment]
Looks promising Claude is doing a good job ;-):

Let it run now

Fix Summary
Layer 1: Async IOReport with Timeout ([apple_silicon_gpu.cpp](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/osx/apple_silicon_gpu.cpp))
Added ioreport_create_samples_with_timeout() wrapper using GCD dispatch queue
2-second timeout prevents indefinite blocking on IOReportCreateSamples()
Returns nullptr on timeout, triggering cached fallback
Layer 2: Cached Metrics Fallback ([apple_silicon_gpu.cpp](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/osx/apple_silicon_gpu.cpp))
Added CachedGpuMetrics structure storing last successful metrics
Cache valid for 5 seconds
On timeout, returns cached values to maintain display continuity
Layer 3: Graceful Skip ([btop.cpp](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L793-834))
First 2 stalls: Warning logged, cycle skipped (no restart)
Only after 3 consecutive stalls: Thread restart
Prevents unnecessary restarts under temporary system load
Layer 4: Adaptive Timeout ([btop.cpp](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L451-475))
Base timeout: 5 seconds
Increases +2s per recent stall (up to 15s max)
On macOS: Further increases if load average >4.0 or >8.0
The build is successful. You can run ./bin/btop to test the fix. The log file at ~/.local/state/btop.log will now show:

Warning for graceful skips: "Runner thread stall detected (1/3) - skipping cycle"
Error only when restarting: "Runner thread stall (3/3) - restarting thread!"

Ciao Marco

**Comment by marcoxyz123** [comment]
I was running a deeper analyses to find further bugs for the stall behaviour:

## Deep Analysis Complete
I've conducted a thorough investigation of the btop threading issues. Here's my findings:

### 🔴 ROOT CAUSE: Mutex Deadlock After Thread Restart
**Location**: ~[btop.cpp:512](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L512)~ and ~[btop.cpp:817](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L817)~
The critical bug is in the thread restart mechanism:
1 The runner thread holds pthread_mutex_t mtx via RAII thread_lock class
2 When a stall is detected, pthread_cancel() is called
**3** **pthread_cancel** **does NOT guarantee C++ destructor execution** - the mutex remains **LOCKED**
4 A new thread is created, which tries to acquire the same mutex
**5** **DEADLOCK**: The new thread blocks forever on the locked mutex

⠀There are **NO** pthread_cleanup_push/pthread_cleanup_pop handlers to release the mutex on cancellation.

### 🟡 Contributing Issues
| **Bug** | **Location** | **Description** |
|:-:|:-:|:-:|
| **Semaphore No Timeout** | ~[btop.cpp:522](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L522)~ | do_work.acquire() blocks indefinitely - should use try_acquire_for() |
| **macOS Exit Hang** | ~[btop.cpp:212](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L212)~ | pthread_join without timeout on macOS - exit hangs if thread stuck |
| **IOReport Memory Leak** | ~[apple_silicon_gpu.cpp:105-131](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/osx/apple_silicon_gpu.cpp#L105)~ | On timeout, async block may complete after return, leaking CFDictionaryRef |

### Timeline Explanation
Your logs show:
* **02:19:45** - "Stall in Runner thread, restarting!" → Thread cancelled, mutex locked
* **07:09:54** - "Failed to join _runner thread on exit!" → ~5 hours later, trying to exit

⠀The system ran in a **degraded/frozen state** for 5 hours because the restarted thread was deadlocked on the mutex.

### Recommended Fixes
**Best Solution**: Replace pthread_cancel with cooperative termination:

// In thread loop:
while (not Global::quitting) {
	// Use timed wait instead of blocking acquire
	if (not do_work.try_acquire_for(std::chrono::seconds(5))) {
		// Check if we should terminate
		if (should_terminate.load()) break;
		continue;
	}
	// ... rest of loop
}
**Alternative**: Add pthread cleanup handlers:

static void cleanup_mutex(void* arg) {
	pthread_mutex_unlock((pthread_mutex_t*)arg);
}

// In _runner():
pthread_cleanup_push(cleanup_mutex, &mtx);
pthread_mutex_lock(&mtx);
// ... thread code ...
pthread_cleanup_pop(1);

Implementing this to

Ciao Marco

**Comment by marcoxyz123** [comment]
In Additional I advised Claude to do a bigger bug analyses for the overall code with my workflows here are the findings (I checked them it is true ;-) ):

# ### New Bugs Found
**🔴 BUG 1: Linux NVML Multi-GPU Thread Leak (HIGH)
File**: ~[src/linux/btop_collect.cpp:1289-1487](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/linux/btop_collect.cpp#L1289-L1487)~
**Issue**: PCIe throughput threads leak when system has multiple NVIDIA GPUs.

// Line 1289 - threads declared OUTSIDE loop
std::thread pcie_tx_thread, pcie_rx_thread;

for (unsigned int i = 0; i < device_count; ++i) {
	// Line 1334-1349 - threads created for EACH GPU
	if (gpus_slice[i].supported_functions.pcie_txrx ...) {
		pcie_tx_thread = std::thread([...]() {...});  // Overwrites previous!
		pcie_rx_thread = std::thread([...]() {...});  // Overwrites previous!
	}
	// ...
	// Line 1481-1487 - Only joins LAST iteration's threads
	if (...) {
		pcie_tx_thread.join();
		pcie_rx_thread.join();
	}
}
**Impact**: For multi-GPU systems, threads from GPUs 0..N-1 are never joined, causing:
* Thread resource leaks
* Undefined behavior (joinable threads destroyed)
* Gradual system resource exhaustion

⠀**Fix**: Move thread variables inside loop OR use a vector of threads.

**🔴 BUG 2: Intel GPU Shutdown Not Called (HIGH)
File**: ~[src/btop.cpp:251-254](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/btop.cpp#L251-L254)~
**Issue**: Intel GPU cleanup is missing from exit handler.

// clean_quit() only calls:
#ifdef GPU_SUPPORT
	Gpu::Nvml::shutdown();
	Gpu::Rsmi::shutdown();
	// MISSING: Gpu::Intel::shutdown();  
#endif
**Impact**: When btop exits with Intel GPU:
* engines structure not freed (~[src/linux/btop_collect.cpp:1854-1856](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/linux/btop_collect.cpp#L1854-L1856)~)
* Perf counters may remain open
* Memory leak

⠀**Fix**: Add Gpu::Intel::shutdown(); to clean_quit().

**🟡 BUG 3: NetBSD Memory Leaks (MEDIUM)
File**: ~[src/netbsd/btop_collect.cpp](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/netbsd/btop_collect.cpp)~
**Issue**: Multiple early return paths leak prop_dictionary_t objects.

// update_sensors() - Lines 287-376
prop_dictionary_recv_ioctl(fd, ..., &dict);  // dict allocated
close(fd);

if (prop_dictionary_count(dict) == 0) {
	Logger::warning(...);
	return;  // LEAK: dict not released
}

if (prop_object_type(fields_array) != PROP_TYPE_ARRAY) {
	return;  // LEAK: dict not released
}

if (regcomp(&r, ...) != 0) {
	return;  // LEAK: dict and fields_iter not released
}
Same pattern in get_battery() (lines 440-474).
**Impact**: Memory leaks accumulate on sensor/battery read errors.
**Fix**: Use RAII wrapper or ensure all return paths release resources.

**🟡 BUG 4: Unsafe strcpy in Intel GPU Code (MEDIUM)
File**: ~[src/linux/intel_gpu_top/intel_name_lookup_shim.c:88,90,322](vscode-webview://00q330esbc7ln22utds613l7f21bjqbfsjfsd45cndokn5vabq3p/src/linux/intel_gpu_top/intel_name_lookup_shim.c#L88)~
**Issue**: strcpy without bounds checking:

// Line 88-90
strcpy(dev_name, "(unknown)");
strcpy(dev_name, info->codename);

// Line 322
strcpy(buf, dent->d_name);
**Impact**: Potential buffer overflow if source exceeds destination buffer.
**Fix**: Replace with strncpy or snprintf.

**🟢 BUG 5: Silent Exception Swallowing (LOW)
Multiple Files**
**Issue**: Empty catch blocks hide errors:

// src/linux/btop_collect.cpp:538, 761, 801, 837, etc.
catch (...) {}

// src/freebsd/btop_collect.cpp:333
catch (...) {}

// src/osx/btop_collect.cpp:846
catch (...) {}
**Impact**: Bugs are silently swallowed, making debugging difficult.
**Fix**: Log exceptions or handle them meaningfully.

### Summary Table
| **Bug** | **Severity** | **Platform** | **File** | **Fix Complexity** |
|:-:|:-:|:-:|:-:|:-:|
| Multi-GPU Thread Leak | 🔴 HIGH | Linux | btop_collect.cpp | Easy |
| Intel GPU Shutdown | 🔴 HIGH | Linux | btop.cpp | Trivial |
| NetBSD Memory Leaks | 🟡 MEDIUM | NetBSD | btop_collect.cpp | Moderate |
| Unsafe strcpy | 🟡 MEDIUM | Linux | intel_name_lookup_shim.c | Easy |
| Silent Exceptions | 🟢 LOW | All | Multiple | Easy |


Ciao Marco

**Comment by marcoxyz123** [comment]
So done Number 2 was also affecting macOS x with intel GPU because it was never compiled since the fiction was defined in linux.

Bug Fixes Completed
1. Linux NVML Multi-GPU Thread Leak (HIGH)
File: btop_collect.cpp:1289-1487

Issue: PCIe threads were created in a loop but overwritten without joining, causing thread leaks on multi-GPU systems.

Fix: Added pcie_threads_created flag to track thread creation and use consistent conditions for create/join.

2. Intel GPU Shutdown Not Called (HIGH)
Files: btop.cpp:254, btop_shared.hpp:215-217, osx/btop_collect.cpp:532-534

Issue: Gpu::Intel::shutdown() was never called in clean_quit(), leaking Intel GPU resources.

Fix: Added shutdown call and required declarations/stubs for cross-platform compatibility.

3. NetBSD Memory Leaks (MEDIUM)
File: netbsd/btop_collect.cpp

Issue: prop_dictionary and prop_object_iterator resources leaked on early returns in update_sensors() and get_battery().

Fix: Added RAII-style cleanup lambdas that execute on all exit paths.

4. Unsafe strcpy in Intel GPU Code (MEDIUM)
File: intel_name_lookup_shim.c:81-97

Issue: Used strcpy() without bounds checking, risking buffer overflow.

Fix: Replaced with snprintf() with proper size limits.

5. Silent Exception Swallowing (LOW)
File: btop_collect.cpp

Issue: Empty catch blocks made debugging difficult.

Ciao Marco

**Comment by marcoxyz123** [comment]
And the list goes on:

Deep Analysis Complete: btop++ Long-Running Stability Issues
I've completed a comprehensive analysis across all platforms. Here's a prioritized summary of 35+ issues that could affect long-running stability:

🔴 CRITICAL ISSUES (Fix Immediately)
1. Dispatch Semaphore Leak - macOS
apple_silicon_gpu.cpp:132

dispatch_semaphore_t created but never released on success path
Impact: ~864,000 leaked semaphores/day at 10Hz → 1.7GB+ Mach port exhaustion
Fix: Add dispatch_release(sem); before return at line 162
2. Race Conditions on Shared Power/Temperature Variables
btop_shared.hpp:112-120

cpuPower, gpuPower, anePower, cpuTemp, gpuTemp are NOT atomic
Written by GPU thread, read by draw thread without synchronization
Impact: Data corruption, torn reads, inconsistent metrics
3. Uninitialized CFDictionary Release - macOS
btop_collect.cpp:1102-1132


CFDictionaryRef properties;  // UNINITIALIZED
IORegistryEntryCreateCFProperties(...);
if (properties) { ... }
CFRelease(properties);  // Called even if never set!
Impact: Crash or memory corruption during disk enumeration
4. File Descriptor Leak in Process Collection - Linux
btop_collect.cpp:3040-3075

pread.open() then continue without pread.close() in multiple places
Impact: Hundreds/thousands of leaked FDs over 24 hours → eventual failure
5. Unsafe GPU Vector Access
btop_shared.hpp:207

Gpu::gpus vector accessed without synchronization from multiple threads
Vector resizes during reads cause iterator invalidation
Impact: Crashes, undefined behavior
🟠 HIGH PRIORITY ISSUES
6. Spin-Lock Busy-Wait Starvation
btop_tools.cpp:513-514


while (atom.load() == old) busy_wait();  // 100% CPU spin
Impact: Performance degradation, battery drain, thread starvation
7. Thread Restart Race Condition
btop.cpp:844-859

should_terminate set before thread_trigger() - signal can be missed
Impact: Lost thread wakeup, runner thread never restarts
8. IOHIDEventSystemClientRef Leaks on Error Paths - macOS
apple_silicon_gpu.cpp:485-544

If services == nullptr, function returns without releasing system and matching
Impact: Kernel connection leak on sensor enumeration failures
9. Unbounded disks_stats_promises Map - Linux
btop_collect.cpp:2151

Entries added for each mountpoint, never erased when unmounted
Impact: Memory growth on systems with dynamic disk mounting
10. ZFS Exception Path File Descriptor Leak - Linux
btop_collect.cpp:2551-2552

Exception handler continues without closing diskread
Impact: FD leak per ZFS error
11. Unchecked popen() in FreeBSD
btop_collect.cpp:514-515

popen() return value not checked before use
Impact: Crash when using fgets on nullptr
12. Config Lock Scope Too Narrow
btop.cpp:929-943

Config could be modified between lock and thread_trigger
Impact: Thread uses stale/torn configuration
🟡 MEDIUM PRIORITY ISSUES
13. sub_channels CFMutableDictionaryRef Leak - macOS
apple_silicon_gpu.cpp:404-405

Populated by IOReportCreateSubscription but never released
14. Unbounded ignore_list Vector - Linux
btop_collect.cpp:2141, 2309

Failed disk mountpoints accumulate indefinitely
15. Dead Process Set Growth When Paused - Linux
btop_collect.cpp:2818, 3191-3210

dead_procs set never cleared in pause mode → 10,000+ PIDs accumulate
16. GPU Percent Deques Width-Dependent Cleanup - macOS
btop_collect.cpp:443-499

Push unconditional, cleanup only when width != 0
Impact: Unbounded growth during startup
17. Async IOReport Sample Race Condition - macOS
apple_silicon_gpu.cpp:140-144

Multiple timeouts can leak CFDictionary objects
18. Missing Unlock in Error Path
btop.cpp:546-551

Mutex locked but thread continues with corrupted state
19. isWhole() Crashes on nullptr - macOS
btop_collect.cpp:1052-1057

CFBooleanGetValue(isWhole) with nullptr crashes
20. Thread Join Without Joinability Check - Linux
btop_collect.cpp:1503-1506

PCIe threads joined without checking joinable()
Summary Table by Platform
Platform	Critical	High	Medium	Total
macOS	3	4	5	12
Linux	2	4	4	10
FreeBSD	1	1	1	3
Cross-platform	2	3	3	8
Total	8	12	13	33
Recommended Fix Order
Phase 1 - Immediate (Crash/Resource Exhaustion):

Dispatch semaphore leak (#1)
Uninitialized CFDictionary release (#3)
Linux file descriptor leaks (#4)
Make shared variables atomic (#2)
Phase 2 - High Priority (Thread Safety):
5. Protect GPU vector with mutex (#5)
6. Replace spin-locks with condition variables (#6)
7. Fix thread restart race (#7)
8. Fix Config lock scope (#12)

Phase 3 - Stability Hardening:
9. Fix all IOKit error path leaks
10. Add cleanup for Linux static containers
11. Add bounds checking before .at() calls

Fix this to.

**Comment by deckstose** [comment]
Implementation details are better discussed in a merge request

> 2026-01-13 (02:19:45) | ERROR: Stall in Runner thread, restarting!
2026-01-13 (07:09:54) | WARNING: Failed to join _runner thread on exit!

This was reported somewhere already IIRC

**Comment by marcoxyz123** [comment]
Got it.

**Comment by Ruaneri-Portela** [comment]
I cleaned up the code I wrote and left only the GPU implementation. I made a PR to add it to the main code.

Then we'll add ANE, Power, and GPU usage per process later on.

**Comment by marcoxyz123** [comment]
Gpu added in Processes works for intel to, also Fan speed in Power for intel and AS.. Fixed today all the XXX Bugs in top. is now running since 10 hours seamlessly without any crash. I guess I found also a fix for the scrambled text sometime. Make tomorrow a for on Github and an PR, I guess we only need to replace ANE with Your VRAM Display for Intel Macs on the GPU/CPU panel.

<img width="3236" height="2222" alt="Image" src="https://github.com/user-attachments/assets/42c4b9ef-e9fb-4c65-b570-a84155d4809a" />

**Comment by marcoxyz123** [comment]
[btop-changes-report.md](https://github.com/user-attachments/files/24601845/btop-changes-report.md)

Here the report from the last 3 Days. I will do a merge request from this. The report is not comprehensive I have to go thru this tomorrow.

**Comment by marcoxyz123** [comment]
@Ruaneri-Portela Maybe it is a good Idea I give you access to my gitlab server than you can add your intel GPU stuff right there and we make than the PR?

**Comment by Ruaneri-Portela** [comment]
> Gpu added in Processes works for intel to, also Fan speed in Power for intel and AS.. Fixed today all the XXX Bugs in top. is now running since 10 hours seamlessly without any crash. I guess I found also a fix for the scrambled text sometime. Make tomorrow a for on Github and an PR, I guess we only need to replace ANE with Your VRAM Display for Intel Macs on the GPU/CPU panel.
> 
> <img alt="Image" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/95569770/535338909-42c4b9ef-e9fb-4c65-b570-a84155d4809a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgzNTgxNTcsIm5iZiI6MTc2ODM1Nzg1NywicGF0aCI6Ii85NTU2OTc3MC81MzUzMzg5MDktNDJjNGI5ZWYtZTlmYi00YzY1LWI1NzAtYTg0MTU1ZDQ4MDlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTE0VDAyMzA1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3OTI2ZThlZmI4NGMwZGM3MDIxY2I3OWI3MmVkMzdjMWY1MGRlY2EzZWY2YThhZjYxYTNmODkwNzM2YmRiNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pe6y3TDmme3byFCP7ISI0LhVLR0YzkOb74pULLekGbg">

Look, that's exactly what I wanted.

It turned out great.

When I refer to memory, I mean both dedicated and integrated GPUs. On the Mac M, even if the memory is shared, the Mac shows how much RAM is allocated to VRAM.

If it's dedicated, it shows how much internal memory is available on the drive. Although, as I mentioned, I can't test this because I don't have a Mac in that configuration.

**Comment by Ruaneri-Portela** [comment]
If you don't mind @marcoxyz123, shall we split the implementation of ANE and power usage into another PR?

**Comment by marcoxyz123** [comment]
Let me check this with the VRAM

**Comment by marcoxyz123** [comment]
@Ruaneri-Portela Let us see where end up. All together there a massive changes in the source code of btop overall. I implement now the VRAM stuff (found it also for M(N))

**Comment by marcoxyz123** [comment]
@Ruaneri-Portela 

So finally:

CPU only

<img width="2632" height="416" alt="Image" src="https://github.com/user-attachments/assets/96f74e9e-7078-4c67-ac9f-ffc169e2a490" />

CPU & GPU

<img width="2642" height="576" alt="Image" src="https://github.com/user-attachments/assets/58b86ab0-8de9-45e8-8aac-df5858108d5e" />

Killer feature -> Set the VRam allocation for Apple Silicon (Shift + A)

<img width="2616" height="1310" alt="Image" src="https://github.com/user-attachments/assets/ab78d71e-2349-4a9b-a78f-a8ea8e0909a4" />

<img width="2702" height="1430" alt="Image" src="https://github.com/user-attachments/assets/b84f408f-f728-4da8-bca7-915ce1540b2c" />

MEM Cleanup for minimal view (13 Lines) Bar view (Shift + B ) Meter view (Shift + M). Vram in the MEM panel Cycle with Free (Shift + V) (Vram only -> Vram + Free -> Free only)

<img width="1174" height="576" alt="Image" src="https://github.com/user-attachments/assets/fd504465-2d39-432d-947e-f34614353085" />

<img width="1202" height="584" alt="Image" src="https://github.com/user-attachments/assets/b1d72611-774c-4d2c-9e97-715e2b8e3286" />

<img width="2638" height="1318" alt="Image" src="https://github.com/user-attachments/assets/05e9c0cf-dc36-40f2-95eb-adc5ed5a4471" />

Enhancment and Copy feature of PROC, Hide Command for better visibility (Shift + C), Sort by Gpu, Click on PID -> Copy to Clipboard, Click on Command -> Copy to clipboard 

<img width="2642" height="1340" alt="Image" src="https://github.com/user-attachments/assets/4c512631-a465-498e-920c-2e5e1c8678a9" />

I do a last check for bugs and push it from my Gitlab to Github

Ciao Marco

**Comment by marcoxyz123** [comment]
Love it ;-)

<img width="2646" height="1744" alt="Image" src="https://github.com/user-attachments/assets/5d13a503-d87c-4f68-91ee-33c7c9a32b41" />

**Comment by marcoxyz123** [comment]
@emanuelez  @Ruaneri-Portela 

PR: https://github.com/aristocratos/btop/pull/1474

Fork: https://github.com/marcoxyz123/btop

Branch: apple-silicon-comprehensive-monitoring

Let me know for gaps to address with Apple Intel.

**Comment by aristocratos** [comment]
Read [CONTRIBUTING.md](https://github.com/aristocratos/btop/blob/main/CONTRIBUTING.md)

> Submissions where the majority of the code is AI generated must be marked with [AI generated].
> 
> "Vibe coded" PR's where it seems like the author doesn't understand the generated code will be dismissed.

Hint, an +10000 lines PR out of the blue is likely never gonna get merged, regardless if it's AI generated or not.

**Comment by achille** [comment]
here's a related pr: https://github.com/aristocratos/btop/pull/1505
if folks with M3, M4 an M5 macs could test, I'd really appreciate it

for my specific needs I'm focused on energy efficiency and need to track the e-cores
but it can be further extended to get gpu metrics

it's pulling  metrics w/out sudo via reverse engineered IOReport data
build here: https://github.com/achille/btop/releases/tag/v1.4.1-pe-cores-preview

@aristocratos this only adds ~400 LoC for ioreport metric pulling; the rest is drawing & unit tests

**Comment by gabimoncha** [comment]
doesn't seem to work on M1 Max, pressing 5 and nothing happens

even reinstalled btop with brew again and still nothing

**Comment by hermeschen1116** [comment]
> doesn't seem to work on M1 Max, pressing 5 and nothing happens
> 
> even reinstalled btop with brew again and still nothing

I think the feature doesn't official release yet. It just being merged into main.

**Comment by gdibble** [comment]
I'm excited to see GPU stats for my M4 Max Mac Studio. Any news on this merging to trunk?

**Comment by ohelabs** [comment]
Same as @gdibble, looking for when this will land in a release?

**Comment by Grazulex** [comment]
I'm using LLM on my laptop...will be great if we can have this type of view too on Mac Pro M5 !!
Noo, I confirme that the feature is not actif on this laptop (Press 5 don't show the GPU graph)

**Comment by LetsGoDoItAll** [comment]
> I'm excited to see GPU stats for my M4 Max Mac Studio. Any news on this merging to trunk?

Not spam, just showing interest

**Comment by gdibble** [comment]
> > I'm excited to see GPU stats for my M4 Max Mac Studio. Any news on this merging to trunk?
> 
> Not spam, just showing interest 

True

**Comment by martinoturrina** [comment]
Thank you guys for the work! this looks so cool!
